pkg/coscheduling/coscheduler.go

```diff
@@ -0,0 +1,484 @@
+			metav1.GetOptions{})
+		if err != nil {
+			klog.Error(err)
+			return
+		}
+		pgCopy := pg.DeepCopy()
+		pgCopy.Status.ScheduleStartTime = metav1.Now()
+
+		var patch []byte
+		patch, err = util.CreateMergePatch(pg, pgCopy)
+		if err != nil {
+			return
+		}
+
+		pg, err = cs.client.SchedulingV1alpha1().PodGroups(pg.Namespace).Patch(context.TODO(), pg.Name,
+			types.MergePatchType, patch, metav1.PatchOptions{})
+		if err != nil {
+			return
+		}
+	}
+
+	// Start coscheduling
+	klog.V(4).Infof("Start scheduling %v", fullName)
+	pendingPods := cs.operation.GetPodNodePairs(fullName)
+	if pendingPods == nil {
+		klog.V(4).Infof("Can not found pending pods for %v", fullName)
+		return
+	}
+	pendingPodNameIDs := cs.operation.GetPodNameUIDs(fullName)
+	if pendingPodNameIDs == nil {
+		klog.V(4).Infof("Can not found pending IDs for %v", fullName)
+		return
+	}
+	pendingPodsMap := pendingPods.Items()
+	if uint32(len(pendingPodsMap)) < pgs.PodGroup.Spec.MinMember-pgs.PodGroup.Status.Scheduled {
+		return
+	}
+
+	// firtstPodUid just record the first pod int the map. When we delete any one of the pending pods,
+	// it would trigger evict of all cache.
+	firstPodUid := ""
+	for uid, pair := range pendingPodsMap {
+		waitingPod := cs.frameworkHandler.GetWaitingPod(types.UID(uid))
+		if waitingPod == nil {
+			// to avoid sig send, but scheduler cache have not been flushed
+			klog.V(4).Infof("Remove pod uid %v, pair %v, beacause waitingPod is nil", uid, pair.Object)
+			pendingPods.Delete(uid)
+			return
+		}
+		if firstPodUid == "" {
+			firstPodUid = uid
+		}
+		waitingPod.Allow(Name)
+		pendingPods.Delete(firstPodUid)
+		klog.V(4).Infof("Allowed pod %+v", pair.Object)
+	}
+}
+
+// rejectPod rejects pod in cache
+func (cs *coSchedulingPlugin) rejectPod(uid types.UID) {
+	waitingPod := cs.frameworkHandler.GetWaitingPod(uid)
+	if waitingPod == nil {
+		return
+	}
+	waitingPod.Reject("Group failed")
+}
+
+// ReconcileStatus reconcile pod cache states and decide when to schedule
+func (cs *coSchedulingPlugin) ReconcileStatus(stopChan <-chan struct{}) {
+	for {
+		select {
+		case pgName := <-cs.startChan:
+			cs.UpdateBatchCache()
+			cs.StartBatchSchedule(pgName)
+		case <-stopChan:
+			klog.Info("Reconcile exit")
+			return
+		}
+	}
+}
+
+// sendStartScheduleSignal send stat scheduling signal to scheduler
+func (cs *coSchedulingPlugin) sendStartScheduleSignal(pgName string) {
+	cs.startChan <- pgName
+	klog.Infof("Send StartScheduleSignal for %v success", pgName)
+}
+
+// pgAdded reacts to a PG creation
+func (cs *coSchedulingPlugin) pgAdded(obj interface{}) {
+	key, err := cache.MetaNamespaceKeyFunc(obj)
+	if err != nil {
+		klog.Error(err)
+		return
+	}
+	pg := obj.(*v1alpha1.PodGroup)
+	if pg.Status.Phase == v1alpha1.PodGroupFinished || pg.Status.Phase == v1alpha1.PodGroupFailed {
+		return
+	}
+	pgsObj := cs.cache.Get(key)
+	if pgsObj == nil {
```

Comment:
It seems like `cs.cache` is the source of truth to get the canonial state of the PodGroup. However, it seems when an existing PodGroup event comes in, it's not refreshed in `cs.cache`? Would that result in some inconsistency?

Reply:
> It seems like `cs.cache` is the source of truth to get the canonial state of the PodGroup. However, it seems when an existing PodGroup event comes in, it's not refreshed in `cs.cache`? Would that result in some inconsistency?

Yes , I would add update logic. So we do not need get it from apiserver as you  commented 
https://github.com/kubernetes-sigs/scheduler-plugins/pull/57#discussion_r493117929